File: SortingAlgorithms/QuickSort/quick.py
import random as rand
import os

#increments the incleft and incright values until it finds a value greater than and
#less than the partition respectively
def increment(nums,incleft,incright,left,right):
    while incleft<right and nums[incleft] < nums[right]: incleft+=1
    while incright>left and nums[incright] > nums[right]: incright-=1
    return incleft,incright

#partitions the array between the ranges such everything left of the partition is less
#and everything greater than the partition is greater than it
def partition(nums,left,right):
    pI = rand.randint(left,right) #chooses random partition
    nums[pI],nums[right] = nums[right],nums[pI] #move partition to the back
    incLeft,incRight = increment(nums,left,right-1,left,right) #increment to find first inc values
    #iterates until the inc values cross each other which indicates that the partition can be put back
    while incLeft < incRight:
        nums[incLeft],nums[incRight] = nums[incRight],nums[incLeft]
        incLeft,incRight = increment(nums,incLeft,incRight,left,right)
    #puts the partition back in its spot
    nums[right],nums[incLeft] = nums[incLeft],nums[right]
    return incLeft #returns new position of the partition

#quick sorts nums
def quick(nums,left=0,right=None):
    if(right is None): right = len(nums-1)
    if(left<right): #until the partition is 1 long
        pI = partition(nums,left,right) #partitions and gets index of partiton
        quick(nums,left,pI-1) #divide and conquers recursively
        quick(nums,pI+1,right)

#forms the number array from file
filePath = os.path.dirname(os.getcwd())+"/testinput.txt"
inputData = open(filePath,'r').read()
nums = inputData.strip().split()
nums = [int(num) for num in nums]

#sorts
quick(nums)
print(nums)

File: SortingAlgorithms/QuickSort/quick.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <string>
#include <time.h>
#include <algorithm>

using namespace std;

/*
This function increments the left and right to find the new left and right
*/
void increment(const vector<int> &nums,int &left,int &right,const int partValue){
    while(nums[left]<partValue)
        ++left;
    while(nums[right]>partValue)
        --right;
}

int partition(vector<int> &nums,int leftInd, int rightInd){
    //Chooses partition index (ideally would get sample median or other method)
    size_t partInd = rand() % (rightInd-leftInd) + leftInd; 
    int partValue = nums[partInd];

    //Moves partition var to the rightmost spot
    swap(nums[partInd],nums[rightInd]);

    //finds the first var greater than the partition from the left and vice-verse for the right
    int left = leftInd, right = rightInd-1;
    increment(nums,left,right,partValue);

    //iterates until left>=right at which point every value less than the partition will be to its left
    while(left<right){
        swap(nums[left],nums[right]);
        increment(nums,left,right,partValue);
    }

    //move the partition var back to its spot
    swap(nums[left],nums[rightInd]);

    return left; //return the index of the partition
}

void quick(vector<int> &nums,int left,int right){
    if(left<right){
        //gets the partion index from the partitioned vector
        int partInd = partition(nums,left,right);

        //Recursively partition/sort left and the right of the partition
        quick(nums,left,partInd-1);
        quick(nums,partInd+1,right);
    }
}

int main(){
    ifstream unsortedData ("../testinput.txt");
    vector<int> unsortedInts;
    
    //populate vector with input data
    int temp;
    while(unsortedData>>temp) unsortedInts.push_back(temp);
    unsortedData.close();

    //sort vector
    srand (time(NULL));
    quick(unsortedInts,0,unsortedInts.size()-1);

    //print vector
    for(int num:unsortedInts)
        cout << num << " ";
    cout<<"\n";
}